### PR TITLE
Fix: Don't show screenshot GUI in screenshots

### PR DESCRIPTION
**Pull request by 2TallTyler**
## Motivation / Problem

Screenshots captured using the "Normal screenshot" button show the screenshot GUI. The other screenshot modes do not.

Unnamed, 1920-11-20

I often see new players posting screenshots using this button on Discord, Reddit, etc., which is typically met with suggestions to use another screen capture tool which doesn't show the screenshot GUI.

I can imagine a player wanting to show other GUI elements like vehicle windows, the status bar ("look at how much money I have!"), and even the main toolbar at top. But I don't see a reason to show the screenshot GUI itself!

## Description

Upstreams [a JGRPP commit](https://github.com/JGRennison/OpenTTD-patches/commit/a65be9b7471753603ecc37f3234125fc06c40853) to hide the screenshot GUI when taking a screenshot.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/screenshot_gui.cpp [hunk 1]
```diff
@@ -72,3 +72,18 @@ void ShowScreenshotWindow()
 	CloseWindowById(WC_SCREENSHOT, 0);
 	new ScreenshotWindow(&_screenshot_window_desc);
 }
+
+void SetScreenshotWindowHidden(bool hidden)
+{
+	ScreenshotWindow* scw = (ScreenshotWindow*)FindWindowById(WC_SCREENSHOT, 0);
```
Comment (LordAro) [inline src/screenshot_gui.cpp]:
```suggestion
	ScreenshotWindow *scw = (ScreenshotWindow *)FindWindowById(WC_SCREENSHOT, 0);
```

File: src/screenshot_gui.cpp [hunk 2]
```diff
@@ -72,3 +72,18 @@ void ShowScreenshotWindow()
...
+	if (scw != nullptr) {
+		if (hidden) {
+			scw->SetDirty();
+			SetBit(scw->left, 30);
+		}
+		else {
```
Comment (LordAro) [inline src/screenshot_gui.cpp]:
```suggestion
		} else {
```

File: src/screenshot_gui.cpp [hunk 3]
```diff
@@ -72,3 +72,18 @@ void ShowScreenshotWindow()
...
 	new ScreenshotWindow(&_screenshot_window_desc);
 }
+
+void SetScreenshotWindowHidden(bool hidden)
+{
+	ScreenshotWindow* scw = (ScreenshotWindow*)FindWindowById(WC_SCREENSHOT, 0);
+	if (scw != nullptr) {
+		if (hidden) {
+			scw->SetDirty();
+			SetBit(scw->left, 30);
```
Comment (LordAro) [inline src/screenshot_gui.cpp]: What is 30 here? Do we have a define/constant for it that can be used instead?

File: src/screenshot_gui.cpp [hunk 4]
```diff
@@ -72,3 +72,18 @@ void ShowScreenshotWindow()
 	CloseWindowById(WC_SCREENSHOT, 0);
 	new ScreenshotWindow(&_screenshot_window_desc);
 }
+
+void SetScreenshotWindowHidden(bool hidden)
```
Comment (LordAro) [inline src/screenshot_gui.cpp]: doxygen doc comments, ideally

**Review by LordAro — DISMISSED** [review]
Can't blame JGR for the codestyle issues, you did those all by yourself ;) (In theory you should be able to add JGRPP repo as a remote and just cherry-pick it)

File: src/screenshot_gui.cpp (hunk 3 above)
Comment (glx22) [reply src/screenshot_gui.cpp]: 30 is to make `left` very big so the window is moved out of the main screen.
I think it doesn't need to be a define/constant, but at least a comment to explain why 30.

Comment (frosch123) [reply src/screenshot_gui.cpp]: Maybe it should be something like `scw->left = 2 * _screen.width`. Setting bit 30 is pretty arbitrary.

Comment (2TallTyler) [reply src/screenshot_gui.cpp]: I suppose then the issue becomes remembering the previous location of the window, when it's time to un-hide it. I've added a comment to explain it for now; let me know what you think.

**Comment by 2TallTyler** [comment]
Someday I'll learn not to stay up past my bedtime to submit PRs and miss silly things. 🙄

I prefer not to cherry-pick because going through the motions of adding the code myself helps me understand it and (when I'm paying attention) catch missing code style, Doxygen, or references to other code not present in vanilla.

File: src/screenshot_gui.cpp [hunk 5]
```diff
@@ -72,3 +72,21 @@ void ShowScreenshotWindow()
...
 }
+
+/**
+ * Hides or shows the screenshot window when taking a screenshot.
+ * @param hidden Are we hiding the window or showing it again after the screenshot is taken?
+ */
+void SetScreenshotWindowHidden(bool hidden)
+{
+	ScreenshotWindow *scw = (ScreenshotWindow *)FindWindowById(WC_SCREENSHOT, 0);
+	if (scw != nullptr) {
```
Comment (TrueBrain) [inline src/screenshot_gui.cpp]: `if (scw == nullptr) return` makes for much better readable code. In general, if you have an if-case that spans till the end of your function, it is more readable to invert the statement and return early :)

File: src/screenshot_gui.cpp [hunk 6]
```diff
@@ -72,3 +72,21 @@ void ShowScreenshotWindow()
 	CloseWindowById(WC_SCREENSHOT, 0);
 	new ScreenshotWindow(&_screenshot_window_desc);
 }
+
+/**
+ * Hides or shows the screenshot window when taking a screenshot.
+ * @param hidden Are we hiding the window or showing it again after the screenshot is taken?
```
Comment (TrueBrain) [inline src/screenshot_gui.cpp]: `hidden` confuses me. Does it mean the window was hidden, but you are unhiding it, or does it mean it will become hidden? In general, you want to use active forms for function names and parameters there in. So `hide` would be much more clear.

Similar, the function name is misleading. You are not setting it "hidden". You set the visibility of the window. So `SetScreenshotWindowVisibility` would be a lot more verbose.

File: src/screenshot_gui.cpp [hunk 7]
```diff
@@ -72,3 +72,21 @@ void ShowScreenshotWindow()
...
+ */
+void SetScreenshotWindowHidden(bool hidden)
+{
+	ScreenshotWindow *scw = (ScreenshotWindow *)FindWindowById(WC_SCREENSHOT, 0);
+	if (scw != nullptr) {
+		if (hidden) {
+			scw->SetDirty();
```
Comment (TrueBrain) [inline src/screenshot_gui.cpp]: The content of the if/else is a bit weird. In the first body you do `SetDirty` -> `Change left`, in the second you do `Change left` -> `SetDirty`. From a reader perspective, this suggests the order matter or something is going on. Where in reality the order really doesn't matter.

I would suggest to move the `SetDirty` outside of the `if`. Normally you do it after changing `left` , as that just reads better: change the position, mark it as dirty. The reverse: mark it as dirty, change position, feels .. not natural. But code-wise the order doesn't matter.

File: src/screenshot_gui.cpp [hunk 8]
```diff
@@ -72,3 +72,21 @@ void ShowScreenshotWindow()
...
+			SetBit(scw->left, 30); // Move window out of main screen
+		} else {
+			ClrBit(scw->left, 30); // Return window to original position
```
Comment (TrueBrain) [inline src/screenshot_gui.cpp]: Although I get it is easy and cheap to use setting a bit, it is confusing to the reader. Why bit 30? Why not 31? Why not 29? So many questions :)

It might be better to make a const with some large number, and add/subtract that value. It still is a hack, but I understand why you do it.

There is another approach to take, but I am not sure it is better/worse. As hide/show is always called in a pair, you could use a `static` value to safely store the current `scw->left`, and set the `scw->left` to a known off-screen value when hiding. That in effect gives the same result.

Basically, now a bit in `left` is used to indicate it is hidden or not, which is just a bit nasty. And might hurt us in future refactors. Personally, I would have gone for the `static` method, but I am not sure what other developers think about it. For sure the bitset is just asking for trouble, so a const to add/subtract to `scw->left` sounds more sane to me :)

**Review by TrueBrain — COMMENTED** [review]
This is absolute nitpicking. But as you indicate you would like to learn from these PRs, here is some feedback. Despite this PR being a few lines, there are just some issues with the logic used :)

File: src/screenshot_gui.cpp (hunk 7 above)
Comment (JGRennison) [reply src/screenshot_gui.cpp]: The order does matter, it's not accidental.

Comment (TrueBrain) [reply src/screenshot_gui.cpp]: huh? That .. needs a comment in the code for sure ;)

File: src/screenshot_gui.cpp (hunk 8 above)
Comment (TrueBrain) [reply src/screenshot_gui.cpp]: Oops, missed that @frosch123 also commented on this (but GitHub was hiding that comment from me). His solution would also work fine, as between the hide and show commands, the width of the screen cannot change. So that is also a safe thing to use in this case.

File: src/screenshot_gui.cpp (hunk 7 above)
Comment (JGRennison) [reply src/screenshot_gui.cpp]: In trunk, Window::SetDirty marks the area of the screen which the window is over as dirty. If the window is off-screen, then Window::SetDirty will have no effect.

**Comment by 2TallTyler** [comment]
Thank you for the educational nitpicking! 😄 

Let me know what you think of these changes.

File: src/screenshot_gui.h [hunk 9]
```diff
@@ -11,5 +11,6 @@
 #define SCREENSHOT_GUI_H
 
 void ShowScreenshotWindow();
+void SetScreenshotWindowVisibility(bool hidden);
```
Comment (TrueBrain) [inline src/screenshot_gui.h]: very very very minor: `hide` :)

**Review by TrueBrain — DISMISSED** [review]

File: src/screenshot_gui.cpp [hunk 10]
```diff
@@ -14,6 +14,8 @@
 #include "widgets/screenshot_widget.h"
 #include "table/strings.h"
 
+extern DrawPixelInfo _screen;
```
Comment (michicc) [inline src/screenshot_gui.cpp]: Were you searching the `gfx_func.h` header?

Comment (2TallTyler) [reply src/screenshot_gui.cpp]: No, I needed `_screen.width` which I believe is defined in `gfx.cpp`, however if I #include that lots of other things break. Is there a better way to do this?

Comment (michicc) [reply src/screenshot_gui.cpp]: Line 64 of gfx_func.h is `extern DrawPixelInfo _screen;`

Comment (2TallTyler) [reply src/screenshot_gui.cpp]: Would it be better to `#include gfx_func.h`?

Comment (michicc) [reply src/screenshot_gui.cpp]: IMHO yes, especially in a cpp file. There might be a reason to make an exception inside a header file if it would pull in too much other stuff, but that is not a concern for cpp files.

Comment (2TallTyler) [reply src/screenshot_gui.cpp]: Done, thanks.

**Review by michicc — APPROVED** [review]
